LandSandBoat/server
# Pandemonium Warden: Full Rework

**_I affirm:_**

## What does this pull request do?

Port of Pandemonium Warden from WingsXI after countless hours of research/testing.

- Odd-number phases to match retail as simply a short intermediate phase
- Mob spell/skill list defined in mob's lua per phase for pet and PW
  - This removes need for most familyid conditionals in skill luas
- Astral Flow given retail-accurate delay on spawn
- Phase and pet change logic moved to functions
- Added flavor text to PW spawn/death

Mobskill adjustments:
- Cackle: wasn't coded
- Brainjack: was DD instead of DoT
- Diamondhide: AoE mobskill but was only ever applying stoneskin to self
- Fossilizing breath: only used if target is in front

~~I do have one concern on the phase and pet functions: They don't function (har har) as `local function` since they call eachother. It works as global functions but I'm sure this isn't what is preferred~~ Got it working with a local table of functions

## Steps to test these changes

`!pos 200 33 -140 68`
Spawn by trading Pandemonium key: `!additem 2572` or manually `!spawnmob 17056168`

Fight and bask in his greatness.

Outline of fight:

> Each phase has a starting HP amount. The even-numbered phases start low enough that low-hp mobskills are immediately available and less TP is required to use a mobskill
> Odd-numbered phases are dverger form, and are hard-coded to use cackle, then hellsnap, then change phase. No damage can be inflicted. And pets do not cast spells
> Each phase, PW has the job's respective 2-hour ability at 50% of the starting HP
> Final phase is the "true form" and every 25% uses astral flow, which resummons all pets as well as 8 avatars.
>     Everyone except tank should be prepared to get away when this happens, though there's plenty of time to run away unless you get stun locked by pet spells
>         "All avatars are summoned at once, and with them plus the lamps up, its hard to move your character."
>         "You will probably get locked in place and die from game mechanics alone."
> During phase change:
>     PW is stunned to interrupt any current action (since it's not dying, just disappearing)
>     PL are despawned
>     PW/PL are disappeared and model/animation sub are adjusted
>     PW reappears and stun is removed
>     PL are respawned
>     All PW buffs are wiped
>     All mobskill LUA were mostly cleared of family-specific restrictions to let this single PW LUA handle everything
>         phase change sets skill and spell list for PW and PL
> If full wipe happens, DoT will keep PW from regen, which will keep him in current form. If he regens past his phase HP, he resets to phase 1
> In each even phase, he has access to the respective mobskills of that mob model:
>    - Chariots
>      - Phases 2, 4, 6, 8
> 	 - 10k HP each
> 	 - Pets: Archaic Gears
>    - Gulool Ja Ja
>      - Phase 10
> 	 - 15000 HP
> 	 - Pets: Mamool Ja
>    - Medusa
>      - Phase 12
> 	 - 15000 HP
> 	 - Pets: Lamiae
>    - Gurfurlur the Menacing
>      - Phase 14
> 	 - 15000 HP
> 	 - Pets: Trolls
>    - Khimaira
>      - Phase 16
> 	 - 20000 HP
> 	 - Pets: Puks
>    - Hydra
>      - Phase 18
> 	 - 20000 HP
> 	 - Pets: Dahaks
>    - Cerberus
>      - Phase 20
> 	 - 20000 HP
> 	 - Pets: Bombs
>    - Dvergr
>      - Phase 21
> 	 - 147000 HP
> 	 - Pets: Miniature Dvergr
>
